Proton/Sources/ObjC/PRTextStorage.m

```diff
@@ -138,20 +138,20 @@ - (void)setAttributes:(NSDictionary<NSString *, id> *)attrs range:(NSRange)range
     [self endEditing];
 }
 
-- (void)insertAttachmentInRange:(NSRange)range attachment:(Attachment *_Nonnull)attachment {
-    NSCharacterSet *spacerCharacterSet = attachment.spacerCharacterSet;
-    BOOL hasPrevSpacer = NO;
-    if (range.length + range.location > 0) {
-        NSUInteger characterIndex = range.location == 0 ? 0 : range.location - 1;
-        hasPrevSpacer = [spacerCharacterSet characterIsMember:[self.string characterAtIndex:characterIndex]];
-    }
+- (void)insertAttachmentInRange:(NSRange)range attachment:(NSTextAttachment *_Nonnull)attachment withSpacer: (NSAttributedString *) spacer {
```

Comment:
Spacer is now passed in from call site. This logic only determines if spacer needs to be added.